### ISSUE TITLE
Pytorch Loss Function for in batch negative sampling and training models

### ISSUE THREAD

**Issue by krishanudb** [pytorch/pytorch#49985]
I am working on an embedding model, where there is a BERT model, which takes in text inputs and output a multidimensional vector. The goal of the model is to find similar embeddings (high cosine similarity) for texts which are similar and different embeddings (low cosine similarity) for texts that are dissimilar.

When training in mini-batch mode, the BERT model gives a N*D dimensional output where N is the batch size and D is the output dimension of the BERT model.

Also, I have a target matrix of dimension N*N, which contains 1 in the [i, j] th position if the sentence[i] and sentence[j] are similar in sense and -1 if not.

What I want to do is find the loss/error for the entire batch by finding the cosine similarity of all embeddings in the BERT output and comparing it to the target matrix.

What I did was simply multiply the tensor with its transpose and then take elementwise sigmoid.

scores = torch.matmul(document_embedding, torch.transpose(document_embedding, 0, 1))
scores = torch.sigmoid(scores)

loss = self.bceloss(scores, targets)
But this does not seem to work.

Is there any other way to do this?

P.S. What I want to do is similar to the method on "in batch cross-entropy" described in the paper:
https://arxiv.org/pdf/1811.08008.pdf

**Comment by ezyang** [comment]
Support requests like this are better answered in the forums: https://discuss.pytorch.org/
